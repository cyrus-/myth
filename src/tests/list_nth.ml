let s = 
"type nat =
  | O
  | S of nat

type list =
  | Nil
  | Cons of nat * list

let list_nth : list -> nat -> nat |>
  { [] => ( 0 => 0
          | 1 => 0
          | 2 => 0
          | 3 => 0 )
  | [2] => ( 0 => 2
           | 1 => 0
           | 2 => 0
           | 3 => 0 )
  | [1; 2] => ( 0 => 1
              | 1 => 2
              | 2 => 0
              | 3 => 0 )
  | [1] => ( 0 => 1
           | 1 => 0
           | 2 => 0
           | 3 => 0 )
  | [2; 1] => ( 0 => 2
              | 1 => 1
              | 2 => 0
              | 3 => 0 )
  | [3; 2; 1] => ( 0 => 3
                 | 1 => 2
                 | 2 => 1
                 | 3 => 0 )
  } = ?"
