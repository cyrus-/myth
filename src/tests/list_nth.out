(* Generated Thu Nov 13 00:35:27 2014 *)
let list_nth : list -> nat -> nat =
  let rec f1 (l1:list) : nat -> nat =
    fun (n1:nat) ->
      match n1 with
        | O -> (match l1 with
                  | Nil -> O
                  | Cons (n2, l2) -> n2)
        | S (n2) -> (match l1 with
                       | Nil -> O
                       | Cons (n3, l2) -> f1 l2 n2)
  in
    f1
;;
